Arm Extension Code

TimB-87 opened the pull request.

TimB-87 committed:
Start Coding Arm

TimB-87 committed:
Add Extension To Arm

TimB-87 committed:
Make it move to a distance correctly (mostly)

Made it move to a distance correctly, for the most part (hopefully at least)

TimB-87 committed:
Fixed Errors

Fixed red lines

TimB-87 committed:
Begin implementing PID and adding set-points

TimB-87 committed:
Finish implementing PID and add soft limit

TimB-87 committed:
Robot map for extend arm

TimB-87 committed:
Update map and make small edits

Updated map with a function and fixed some small errors in the code to be able to test on the robot.

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -8,17 +8,22 @@
+    Arm arm = new Arm(map.getArmMap());
     private Drive drive = new Drive(map.getDriveMap());
 
+    private ButtonXboxController controller = new ButtonXboxController(0);
```
Don't we already have this controller on line 19?

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+
+    Data data;
+    ArmMap map;
+    private SmartMotorController extendMotor = new SmartMotorController();
```
Do we still need this?

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+
+    public Arm(ArmMap map) {
+        this.map = map;
+        this.data = new Data();
```
You can combine this with line 23

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+
+    public CommandBase open() {
+        return runOnce(() -> {
+            data.setPoint = 2;
```
2 isn't a valid speed

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+        });
+    }
+
+    private double softLimit(double speed) {
```
Private functions at the bottom, please.
Also, remind us to teach you about Modifiers

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+    //
+    //
+    // Will I need this? I don't think so but I'm not deleting it
+    public CommandBase retract(DoubleSupplier motorSpeed) {
```
Public functions above the overrides please.

Order:

- Data members
- Constructors
- Public functions
- Override functions
- Private and Protected functions

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+    // Will I need this? I don't think so but I'm not deleting it
+    public CommandBase retract(DoubleSupplier motorSpeed) {
+        return run(() -> {
+            data.setPoint = (motorSpeed.getAsDouble());
```
Why are there parenthesis around this?

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+            data.setPoint = (speed);
+        }).runsUntil(() -> {
+            if (distance == data.distanceInches) {
+                extendMotor.stopMotor();
```
`stopMotor` should be `data.setPoint = 0.0`. However, this should be done in the `onEnd` call.

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+                data.setPoint = (-speed);
+            }
+        }).runsUntil(() -> Math.abs(distance - data.distanceInches) < 0.5).onEnd(() -> {
+            extendMotor.stopMotor();
```
`stopMotor` should be `data.setPoint = 0.0`

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+    private SmartMotorController extendMotor = new SmartMotorController();
+    private final double SOFT_MAX_DISTANCE = 400.0;
+    private final double SOFT_MIN_DISTANCE = 20.0;
+    PIDController pid = new PIDController(0.03, 0, 0);
```
`private final`

bot190 replied:
The PID constants should come from the Robot Map similar to `SOFT_MAX_DISTANCE`.

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+package frc.robot.subsystems;
+
+import static edu.wpi.first.wpilibj2.command.Commands.runOnce;
```
You shouldn't need this - `runOnce` is a member of `SmartSubsystemBase`

msoucy commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    private SmartMotorController motor = new SmartMotorController();
+    private final double SOFT_MAX_DISTANCE = 20;
```
Are these used?

JoeJanickiChSh replied:
If these are going to be used either make them public or make getter functions that return these.

JoeJanickiChSh commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+    Data data;
+    ArmMap map;
+    private SmartMotorController extendMotor = new SmartMotorController();
+    private final double SOFT_MAX_DISTANCE = 400.0;
```
Get these from the ArmMap with those unused variables in there.

bot190 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -8,17 +8,22 @@
     private ButtonXboxController driveController = new ButtonXboxController(0);
 
     // $Subsystems$
+    Arm arm = new Arm(map.getArmMap());
```
```suggestion
    private Arm arm = new Arm(map.getArmMap());
```

bot190 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -45,7 +50,7 @@ public void robotInit() {
     @Override
     public void configureButtonBindings() {
-
+        controller.a().onTrue(arm.moveToDistance(200, 0.3));
```
Lets not map things to buttons yet. Instead we should send this command to the smartdashboard.

bot190 commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -7,18 +7,24 @@
 @RobotMapFor("Default")
 public class RobotMap {
     // $Maps$
+    private ArmMap armMap = new ArmMap();
```
Don't need this line if you're creating a new `ArmMap` object in `getArmMap()`

bot190 commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -0,0 +1,58 @@
+
+        @Override
+        public void toLog(LogTable table) {
+            // TODO Auto-generated method stub
```
Don't need these `// TODO` comments if they're done!

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+        return runOnce(() -> {
+            data.setPoint = 2;
+        });
+    }
```
What's this command for?

TimB-87 committed:
Fix some minor errors

(fixed some errors probably not very well)

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,159 @@
+        return cmd("Move Distance").onInitialize(() -> {
+            if (distance >= data.distanceInches) {
+                // extend
+                data.setPoint = softLimit(pid.calculate(distance - data.distanceInches));
```
`calculate` takes the current state and the desired state as separate arguments, you shouldn't be subtracting here.

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,155 @@
+        });
+    }
+
+    public CommandBase movetoLow() {
+        return moveToDistance(Level.LOW.get(), SPEED);
+    }
+
+    public CommandBase movetoMedium() {
+        return moveToDistance(Level.MEDIUM.get(), SPEED);
+    }
+
+    public CommandBase movetoHigh() {
+        return moveToDistance(Level.HIGH.get(), SPEED);
+    }
```
These probably don't need to exist.

msoucy replied:
Or if you do want them:

```java
public CommandBase moveTo(Level level) {
    return moveToDistance(level.get(), SPEED);
}
```

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,155 @@
+        return cmd("Move Distance").onInitialize(() -> {
+            if (distance >= data.distanceInches) {
+                // extend
+                data.setPoint = (speed);
```
This should use `softLimit`

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,155 @@
+        });
+    }
+
+    public CommandBase retractDistance(double distance, double speed) {
+        return cmd("Retract Distance").onInitialize(() -> {
+            data.setPoint = (speed);
+        }).runsUntil(() -> {
+            if (distance == data.distanceInches) {
+                data.setPoint = 0;
+                return true;
+            } else {
+                return false;
+            }
+        });
+    }
```
This probably doesn't need to exist since we have `moveToDistanceTwo`.

msoucy replied:
If that's the case, that one should be renamed...

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,155 @@
+
+public class Arm extends SmartSubsystemBase {
+
+    public double SOFT_MAX_DISTANCE = 400.0;
+    public double SOFT_MIN_DISTANCE = 20.0;
```
These should come from the `ArmMap`.

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,155 @@
+
+public class Arm extends SmartSubsystemBase {
+
+    public double SOFT_MAX_DISTANCE = 400.0;
```
400 inches???? I suspect that won't fit in the robot within the rules.

msoucy replied:
It's bigger on the inside

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public void safeState() {
+
```
The safe state here should be "set the setpoint to 0"

TimB-87 committed:
Fix more errors

JoeJanickiChSh commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,158 @@
+
+public class Arm extends SmartSubsystemBase {
+
+    public double SOFT_MAX_DISTANCE;
```
private final double

JoeJanickiChSh commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,158 @@
+    //
+    //
+    // Will I need this? I don't think so but I'm not deleting it
+    public CommandBase retract(DoubleSupplier motorSpeed) {
```
You probably don't need this since you can just pass a negative number into extend.

JoeJanickiChSh commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    // Do I need this? Matt, Ben, or Joe, help
+    public CommandBase retractDistance(double distance, double speed) {
```
Same thing here you do not need this.

TimB-87 committed:
Fix More Errors

Deleted some stuff and moved the bang bang controller into the main code.

TimB-87 committed:
Fix Minor Errors

msoucy commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -12,8 +12,8 @@ public ArmMap() {
     private SmartMotorController extendMotor = new SmartMotorController();
-    private final double SOFT_MAX_DISTANCE = 20;
-    private final double SOFT_MIN_DISTANCE = 1;
+    public final double SOFT_MAX_DISTANCE = 20;
```
If these can be overridden in derived maps, they should be functions instead

msoucy commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -34,9 +34,9 @@ public static class Data implements LoggableInputs {
         public double[] currentAmps;
         public double[] tempCelcius;
 
+        @Override
```
???

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -78,16 +79,16 @@ public CommandBase moveToDistance(double distance, double speed) {
 
-    public CommandBase movetoLow() {
-        return moveToDistance(Level.LOW.get(), SPEED);
+    public CommandBase moveToLow(Level LOW) {
```
What. This completely defeats the purpose of parameterization

TimB-87 replied:
Fixed it, I'll push later

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -108,6 +109,15 @@ public void periodic() {
     }
 
+    private double softLimit(double speed) {
+        if (data.distanceInches > SOFT_MAX_DISTANCE && speed > 0) {
```
Why are there two conditions that do the same thing?

TimB-87 replied:
huh?

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -108,6 +109,15 @@ public void periodic() {
+        } else if (data.distanceInches < SOFT_MIN_DISTANCE && speed < 0) {
+            data.setPoint = (speed * 0.1);
+        }
+        return speed;
```
Why does this return the original speed?

TimB-87 replied:
Ben told me to

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -117,6 +127,7 @@ public CommandBase retract(DoubleSupplier motorSpeed) {
         });
     }
 
+    // Do I need this? Matt, Ben, or Joe, help
```
:sob:

TimB-87 replied:
Same Matt, same

Commit:
Fixed even more errors

Used Finn's laptop though

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,116 @@
+                // extend
+                data.setPoint = softLimit(pid.calculate(data.distanceInches, distance));
+            }
+        }).runsUntil(() -> Math.abs(distance - data.distanceInches) < 0.5).onEnd(() -> {
```
There's a `pid.atSetpoint()` function that you'll probably find useful

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    public CommandBase moveTo(Level level) {
+        return moveToDistanceBangBang(level.get(), SPEED);
```
Once PID works we'll want to switch this to use that instead

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,116 @@
+        return cmd("Move Distance").onInitialize(() -> {
+            if (distance >= data.distanceInches) {
+                // extend
+                data.setPoint = softLimit(pid.calculate(data.distanceInches, distance));
```
@bot190 are we sure we want `softLimit` around the PID call?

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,116 @@
+            this.angle = angle;
+        }
+
+        public double get() {
```
I'd like this to be called `getLength` please?

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -57,6 +62,7 @@ public void populateDashboard() {
-
+        arm.setDefaultCommand(
+            arm.manual(copilotController::getLeftX));
+            }
```
Formatting is all wonky and played tricks on me

Commit:
Fix more errors to make it work

Fixed to make it work correctly when extending

msoucy commented on build.gradle:
```diff
@@ -1,6 +1,6 @@
 plugins {
     id "java"
-    id "edu.wpi.first.GradleRIO" version "2023.2.1"
+    id "edu.wpi.first.GradleRIO" version "2023.1.1"
```
This feels like a step backwards

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,116 @@
+
+    public CommandBase moveToDistancePID(double distance, double speed) {
+        return cmd("Move Distance").onInitialize(() -> {
+            if (distance >= data.distanceInches) {
```
Is this check necessary here? PID controller should handle this

msoucy commented:
You have merge conflicts - see Ben or Liam for help solving them if you're stuck

JoeJanickiChSh committed:
Test bang bang and PID

TimB-87 committed:
Add velocity check

Add velocity check to make the velocity = 0 (likely will be used at the beginning or end of a match)

msoucy commented on src/main/java/frc/robot/maps/Valkyrie.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Override
+    public void SetupLogging() {
```
You know, if someone were to change the name of this to be capitalized properly (here and in the `RobotMap.java`, I wouldn't be upset...

Lucimoot committed:
Fixed code up a lil

resolved request changes off of the pull request

msoucy commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,123 @@
+    public Data data = new Data();
+    public ArmMap map;
+    public final double SPEED = 0.3;
+    private final double retractSpeed = -0.1;
```
`RETRACT_SPEED`

msoucy commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -0,0 +1,64 @@
+        data.tempCelcius = extendMotor.getTemperatureC();
+    }
+
+    public SmartMotorController getMotor() {
```
Why does this method exist? Is it used anywhere?

msoucy commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public PIDController pid = new PIDController(0, 0, 0);
+    private SmartMotorController extendMotor = new SmartMotorController();
```
Why is just this one `private`?

msoucy commented on src/main/java/frc/robot/maps/subsystems/ArmMap.java:
```diff
@@ -0,0 +1,64 @@
+    public ArmMap() {
+    }
+
+    public PIDController pid = new PIDController(0, 0, 0);
```
These member variables should be set in the constructors, instead of as part of the class members

Lucimoot committed:
Added comments

found some ninja comments on the arm subsystem code and changed the capitalization of some of intake subsystem comments

Lucimoot committed:
resolved some requested changes

did this to make matt happy again

Lucimoot committed:
Undid rebase stuff

removed certain things added by rebasing that Matt told me to remove

bot190 commented on src/main/java/frc/robot/subsystems/Arm.java:
```diff
@@ -0,0 +1,129 @@
+            // Extend
+            data.setPoint = softLimit(map.pid.calculate(data.distanceInches, distance));
+
+        }).runsUntil(() -> Math.abs(distance - data.distanceInches) < 4).onEnd(() -> {
```
This can be simplified to `map.pid.atSetpoint()`.

bot190 reviewed (commented):
One small fix, but this is looking good!

TimB-87 committed:
Fix Tiny Errors

TimB-87 committed:
Set Tolerance for PID
